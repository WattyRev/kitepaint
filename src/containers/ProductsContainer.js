import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { getProductsGrouped } from "../redux/modules/products";
import { getManufacturers } from "../redux/modules/manufacturers";
import { GET_PRODUCTS, GET_MANUFACTURERS } from "../redux/actions";

/**
 * Provides information and actions about/for Designs.
 */
export class ProductsContainer extends React.Component {
  static propTypes = {
    /**
     * A function returning content to be rendered
     */
    children: PropTypes.func.isRequired,
    /**
     * The products indexed by manufacturer id.
     */
    products: PropTypes.objectOf(PropTypes.array.isRequired).isRequired,
    /**
     * A list of manufacturers
     */
    manufacturers: PropTypes.array.isRequired,
    /**
     * A function to trigger the retreival of the products. This should update the redux state,
     * causing products to be provided through redux.
     */
    getProducts: PropTypes.func.isRequired,
    /**
     * A function to trigger the retreival of the manufacturers. This should update the redux state,
     * causing manufacturers to be provided through redux.
     */
    getManufacturers: PropTypes.func.isRequired
  };

  componentDidMount() {
    this.props.getProducts();
    this.props.getManufacturers();
  }

  render() {
    return this.props.children({
      actions: {},
      props: {
        products: this.props.products,
        manufacturers: this.props.manufacturers
      }
    });
  }
}

const mapStateToProps = state => ({
  products: getProductsGrouped(state, "manufacturer"),
  manufacturers: getManufacturers(state)
});

const mapDispatchToProps = {
  getProducts: GET_PRODUCTS,
  getManufacturers: GET_MANUFACTURERS
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ProductsContainer);
