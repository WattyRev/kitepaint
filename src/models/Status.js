import PropTypes from "prop-types";

/**
 * Status can be applied to any model to indicate it's public visibility.
 */
export const statusProp = PropTypes.oneOf(["0", "1", "2"]);
const Status = {
  "0": "Private",
  "1": "Unlisted",
  "2": "Public",
  PRIVATE: "0",
  UNLISTED: "1",
  PUBLIC: "2"
};
export default Status;
