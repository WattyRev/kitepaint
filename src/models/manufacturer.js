import PropTypes from "prop-types";

/**
 * A manufacturer is a company that manufactures kites and owns one or more products.
 */
const manufacturerState = PropTypes.shape({
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  logo: PropTypes.string.isRequired,
  website: PropTypes.string
});

export default manufacturerState;

/**
 * A mock manufacturer used for testing
 * @type {Object}
 */
const getMockManufacturer = () => ({
  id: "abc",
  name: "Krazy Kites",
  logo: "abc.jpg",
  website: "http://krazykites.com"
});
export { getMockManufacturer };
