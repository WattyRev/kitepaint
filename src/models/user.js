import PropTypes from "prop-types";

/**
 * A user is an individual user account. It contains limited details that are available publically.
 */
const userShape = PropTypes.shape({
  username: PropTypes.string.isRequired,
  loginid: PropTypes.string.isRequired
});

export default userShape;

/**
 * A mock user used for testing.
 */
const getMockUser = () => ({
  username: "FlyFucker",
  loginid: "abc"
});

export { getMockUser };
