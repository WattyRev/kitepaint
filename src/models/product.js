import PropTypes from "prop-types";
import { getMockManufacturer } from "./manufacturer";
import Status, { statusProp } from "./status";

/**
 * A product is a kite that can be customized to create a design.
 */
const productShape = PropTypes.shape({
  id: PropTypes.string.isRequired,
  embed: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  notes: PropTypes.arrayOf(PropTypes.string),
  manufacturer: PropTypes.string.isRequired,
  status: statusProp.isRequired,
  url: PropTypes.string,
  colors: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      color: PropTypes.string.isRequired
    })
  ),
  variations: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      svg: PropTypes.string.isRequired
    })
  )
});

export default productShape;

/**
 * A mock product used for testing
 * @type {Object}
 */
const getMockProduct = () => ({
  id: "product-1",
  embed: "krazykites.com",
  name: "Krazy Kite",
  manufacturer: getMockManufacturer().id,
  url: "http://krazykites.com/krazy-kite",
  colors: [
    {
      name: "red",
      color: "#ff0000"
    },
    {
      name: "black",
      color: "#000000"
    }
  ],
  variations: [
    {
      name: "Standard",
      svg: "<div>Kool Kite</div>"
    }
  ],
  status: Status.PUBLIC
});

export { getMockProduct };
