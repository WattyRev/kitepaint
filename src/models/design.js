import PropTypes from "prop-types";
import { getMockProduct } from "./product";
import Status, { statusProp } from "./status";

/**
 * A design is a user-created entity. It represetns a saved design of a particular product.
 */
const designShape = PropTypes.shape({
  id: PropTypes.string.isRequired,
  // A string representing the created date
  created: PropTypes.string.isRequired,
  // A string representing the updated date
  updated: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  // The ID of the cooresponding product
  product: PropTypes.string.isRequired,
  // The ID of the user that created the design
  user: PropTypes.string,
  variations: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      svg: PropTypes.string.isRequired,
      primary: PropTypes.bool.isRequired
    })
  ),
  // The status of the cooresponding product
  productStatus: statusProp.isRequired,
  // The status of the design
  status: statusProp.isRequired
});

export default designShape;

/**
 * A mock design used for testing
 * @type {Object}
 */
const getMockDesign = () => ({
  id: "design-1",
  created: "01/01/2018",
  updated: "01/02/2018",
  name: "picasso",
  product: getMockProduct().id,
  variations: [
    {
      name: "Standard",
      svg: "<div>picasso standard</div>",
      primary: true
    }
  ],
  status: Status.PRIVATE,
  productStatus: Status.PUBLIC
});
export { getMockDesign };
