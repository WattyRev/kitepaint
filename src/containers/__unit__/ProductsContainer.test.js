import React from "react";
import { shallow, mount } from "enzyme";
import { getMockProduct } from "../../models/Product";
import { getMockManufacturer } from "../../models/Manufacturer";
import { ProductsContainer } from "../ProductsContainer";

describe("ProductsContainer", () => {
  let defaultProps;
  beforeEach(() => {
    defaultProps = {
      products: {
        [getMockManufacturer().id]: [getMockProduct()]
      },
      manufacturers: [getMockManufacturer()],
      getProducts: jest.fn(),
      getManufacturers: jest.fn()
    };
  });
  it("renders with the expected data", () => {
    expect.assertions(1);
    const render = jest.fn(() => <div>Test</div>);
    shallow(<ProductsContainer {...defaultProps}>{render}</ProductsContainer>);
    expect(render.mock.calls[0][0]).toEqual({
      props: {
        isLoading: true,
        products: defaultProps.products,
        manufacturers: defaultProps.manufacturers
      }
    });
  });
  it("triggers getProducts when mounted", () => {
    expect.assertions(1);
    mount(
      <ProductsContainer {...defaultProps}>
        {() => <div>Test</div>}
      </ProductsContainer>
    );
    expect(defaultProps.getProducts).toHaveBeenCalled();
  });
  it("triggers getManufacturers when mounted", () => {
    expect.assertions(1);
    mount(
      <ProductsContainer {...defaultProps}>
        {() => <div>Test</div>}
      </ProductsContainer>
    );
    expect(defaultProps.getManufacturers).toHaveBeenCalled();
  });
});
