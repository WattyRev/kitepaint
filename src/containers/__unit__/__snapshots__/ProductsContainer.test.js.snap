// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsContainer renders with the expected data 1`] = `
<div>
  Test
</div>
`;
